### DISCUSSION TITLE
[Kafka Companion] Assertion of produced messages

### DISCUSSION THREAD

**Discussion by Inithron** [smallrye/smallrye-reactive-messaging#2591] [Q&A] [answered]
Hi,

I have a question to the [Kafka Companion](https://smallrye.io/smallrye-reactive-messaging/latest/kafka/test-companion/): Is it possible to assert if the produced messages (with companion) was acknowledged positive or negative by the consumer (the application under test)?

**Comment by ozangunalp** [comment]
Kafka companion works solely on the Kafka level, so negative acknowledgement is not applicable. 

Acknowledgement being an offset commit (of a particular consumer) in Kafka, you can check the offset commit of a consumer group.
For example, the following will give you offset information per topic-partition :

```java
companion.consumerGroups().offsets("my-consumer-group")
```

On the other hand, the Kafka companion `produce` operation yields the list of produced `RecordMetadata` so that you can correlate offsets.

**Reply by Inithron** [reply]
I am new to Kafka and smallrye-reactive-messaging. So if the negative acknowledgement is not applicable on Kafka level, it is at least possible on smallrye-reactive-messaging level. See [nack](https://javadoc.io/doc/io.smallrye.reactive/smallrye-reactive-messaging-api/latest/org/eclipse/microprofile/reactive/messaging/Message.html#nack(java.lang.Throwable)) and [Negative acknowledgement](https://smallrye.io/smallrye-reactive-messaging/4.21.0/concepts/acknowledgement/#negative-acknowledgement).
[Kafka Companion](https://smallrye.io/smallrye-reactive-messaging/latest/kafka/test-companion/) belongs also to smallrye-reactive-messaging. So in theory it should be possible to assert if an acknowledgement was positive or negative. Or did I get it wrong?

**Reply by ozangunalp** [reply]
I understand your confusion, but your assumption is not true. As the documentation states, the Kafka Companion is a separate library to help test Kafka applications. It is not built on reactive messaging concepts and it is not its goal. 

The negative acknowledgement for Kafka messages allows to apply different [failure handling](https://smallrye.io/smallrye-reactive-messaging/latest/kafka/receiving-kafka-records/#failure-management) strategies but it doesn't mean that the message will be flagged as nacked in the Kafka broker.

You can use [in-memory-connector](https://smallrye.io/smallrye-reactive-messaging/latest/concepts/testing/) for writing such advanced test cases. But it'll not test any Kafka concepts, purely the processing method logic.

**Reply by Inithron** [reply]
Just to be sure that my understanding is now right: If I use the failure strategy "ignore" and set a message in my application to "nack", on Kafka level the message will be committed but without the information if it was positive or negative committed?

**Reply by ozangunalp** [reply, marked as answer]
Yes, in Kafka you can't reject or nack messages individually. You've either processed successfully records on the log (topic-partition) up to a certain point (offset) and committed that offset to the consumer group or you've not.

**Reply by Inithron** [reply]
Thanks for clarification.
